### ISSUE TITLE
Set breakpoint using win.webContents.debugger.sendCommand not working

### ISSUE THREAD

**Issue by fijiwebdesign** [electron/electron#8575]
* Electron version:v1.2.7
* Operating system: darwin x64, Mac OSX Yosemite 10.10.5

### Expected behavior

I'm using electron as a node.js debugger that require()'s the node.js app into the electron render process. source: https://github.com/fijiwebdesign/electron-scope

I'm trying to set a breakpoint on the first line of the required node app entry file following the tests from here: 
https://github.com/electron/electron/blob/702352804239f58e5abcd0b96dbd748b68ab0278/spec/api-debugger-spec.js#L77

*Breakpoint should be set* on the first line of js file given. 
However, it can only be set on when attaching to the debugger via: 
`win.webContents.debugger.attach()`
the breakpoint is **not set** when attaching to the debugger process via:
`win.webContents.openDevTools()`

### Actual behavior

No breakpoint is set. 

### How to reproduce

```
      const win = new electron.BrowserWindow({
		width: 600,
		height: 400
	})
        win.webContents.openDevTools()
	win.webContents.debugger.sendCommand(
		'Debugger.setBreakpointByUrl', {
			lineNumber: 0,
			url: './test.js'
		},
		function (err, result){
			console.log('breakpoint set')
			if(err){
				console.error('Error:', err)
			}
			console.log('Result: ', result)
		})

```

I get the console.log result as: `Result:  { breakpointId: './test.js:0:0', locations: [] }`

But the actual breakpoint isn't set. 

full source of setting breakpoint: https://github.com/fijiwebdesign/electron-scope/blob/setBreakpoint/index.js#L51

**Comment by groundwater** [comment]
Thanks @fijiwebdesign will check it out.

Can you try on the latest version of electron, version 1.4.15, available with `npm install electron@latest`.

**Comment by deepak1556** [comment]
@fijiwebdesign at any point only one debugger sessio can be attached to a webContents, this is currently a limitation of chromium https://bugs.chromium.org/p/chromium/issues/detail?id=129539. So `webContents,debugger` session and `openDevtools` are two different debugging sessions and hence breakpoints wont appear.

Aside from that there are a few issues in your chrome RDP usage.

* Any Debugger domain commands should be used only after completion of `Debugger.enable` command. https://chromedevtools.github.io/debugger-protocol-viewer/tot/Debugger/#method-enable

```
function setBreakpoint(win) {
  win.webContents.debugger.sendCommand('Debugger.enable', function () {
    win.webContents.debugger.sendCommand(
      'Debugger.setBreakpointByUrl', {
        lineNumber: 0,
        url: <absolute URL to file> // should be absolute, not relative
      },
      function (err, result){
        console.log('breakpoint set')
        if(err){
          console.error('Error:', err)
        }
        console.log('Result: ', result)
      })
  })
}
```

* In your example, `setBreakPoint` is being called before the `webContents` finishes loading. At this point even though the debugger is attached it has no clue to a file named test.js, since its not available. You can call it after `did-finish-load` event of `webContents`.

**Comment by fijiwebdesign** [comment]
@groundwater I've updated to Electron v1.4.15

@deepak1556 Thanks. It appears `win.webContents.openDevTools()` will attach without errors to the debug session and close the open session of `win.webContents.debugger`. 

I think it should throw an exception just like `win.webContents.debugger.attach()` if there is already a session for `win.webContents.openDevTools()`. That would make it easier to bridge the understanding that it's a separate debug session from both ends. I will try and write a PR for this if my understanding is correct here - unless this is intended behavior. 

For setting the breakpoint there are some issues: 

1) I never got it to work if I first call: `win.webContents.debugger.sendCommand('Debugger.enable', function () {` 

2) The same code seems to work and then not work any more without changes. I suspect it has to do with the devtools persisting state even after closing electron. I can't figure out what changed to have stopped the breakpoints altogether. 

3) Even if breakpoints are set, chrome debugger doesn't process them unless you reload the page. Both manual and reloading with `win.webContents.reload()` will work. This is true even for `debugger;` statements in code. Same as these issues:
https://github.com/electron/electron/issues/7361
https://github.com/electron/electron/issues/4842
And this workaround: https://github.com/JamesKyburz/electron-breakpoint/blob/master/index.js#L14

4) I had to add 'file://' before the URL of script to set the breakpoint. 
```
'Debugger.setBreakpointByUrl', {
				lineNumber: 0,
				url: 'file://' + url
			}
```
Devtools sometimes resolves the file to a domain and `file:///absolute/path/to/script.js` would work. But if it resolved to "no domain" then `/absolute/path/to/script.js` would work. I can't find the correlation between this resolution of domains. 

This is the code that worked at one point: 

```
// debugClient = win.webContents.debugger
// debug = console.log.bind(console)
function setBreakpoint(debugClient, url, callback) {
	url = require('path').resolve(url)
	debug('setting breakpoint', url)
	try {
		debugClient.attach()
	} catch(e) {
		debug('error attaching debugger', e)
	}

	//debugClient.sendCommand('Debugger.enable', (err) => {
		//debug('enabed debugger', err)
	setTimeout(() => {
		debugClient.sendCommand(
			'Debugger.setBreakpointByUrl', {
				lineNumber: 0,
				url: 'file://' + url
			},
			(err, result) => {
				if(err && err.msg){
					console.error('Error:', err)
				}
				debug('Breakpoint: ', result)
				debugClient.detach()
				if (typeof callback == 'function') {
					callback(result)
				}
			})
	}, 500)
	//})
}
```

I will write a set of tests for this and post it here so it is more clear.

**Comment by zcbenz** [comment]
I'm closing this since it is a limitation of the `win.webContents.debugger` API, to make it work with the devtools is another story and we don't have plans for that.

**Comment by tinganho** [comment]
I think this can be opened again. The limitation is fixed https://bugs.chromium.org/p/chromium/issues/detail?id=129539
https://developers.google.com/web/updates/2017/10/devtools-release-notes#multi-client

**Comment by 4eveRS** [comment]
Since this is no longer a issue in chrome since v63. Can it be fixed at Electron?

**Comment by xland** [comment]
I think... They will not read issues that they have already closed
